File: timkiemthongminh-main/project/backend/app/services/news_service.py
"""
Service layer: business logic cho danh sách tin, danh mục, và tìm kiếm
từ khoá (keyword search) — thao tác trên DataFrame lấy từ
`NewsRepository`. KHÔNG đọc lại CSV, KHÔNG chứa route/HTTP logic.
"""

from __future__ import annotations

import math

import pandas as pd

from app.config import DEFAULT_PAGE_LIMIT, MAX_PAGE_LIMIT
from app.database.news_repository import strip_accents_lower
from app.models.news import CategoriesList, CategoryItem, KeywordSearchResult, NewsItem, NewsList


def _clamp_limit(limit: int) -> int:
    return max(1, min(limit, MAX_PAGE_LIMIT))


def _clamp_page(page: int) -> int:
    return max(1, page)


def _row_to_news_item(row: pd.Series) -> NewsItem:
    comments = row["so_binh_luan"]
    return NewsItem(
        id=int(row["id"]),
        title=row["tieu_de"],
        publish_date=row["ngay_dang"],
        author=row["tac_gia"],
        source=row["nguon"],
        url=row["link"],
        comments=None if pd.isna(comments) else int(comments),
        summary=row["summary"],
        image=None,
    )


def _paginate_df(df: pd.DataFrame, page: int, limit: int) -> tuple[pd.DataFrame, int, int]:
    page = _clamp_page(page)
    limit = _clamp_limit(limit)
    start = (page - 1) * limit
    return df.iloc[start : start + limit], page, limit


def get_news_list(df: pd.DataFrame, page: int = 1, limit: int = DEFAULT_PAGE_LIMIT) -> NewsList:
    """Danh sách tin mới nhất, sắp xếp theo ngày đăng giảm dần."""
    sorted_df = df.sort_values(by="ngay_dang", ascending=False, kind="stable")
    page_df, page, limit = _paginate_df(sorted_df, page, limit)
    total = len(df)
    return NewsList(
        total=total,
        page=page,
        limit=limit,
        total_pages=max(1, math.ceil(total / limit)),
        items=[_row_to_news_item(row) for _, row in page_df.iterrows()],
    )


def get_featured_news(df: pd.DataFrame, limit: int = 5) -> NewsList:
    """Tin nổi bật: N tin có nhiều bình luận nhất."""
    limit = _clamp_limit(limit)
    featured_df = df.sort_values(by="so_binh_luan", ascending=False, kind="stable").head(limit)
    total = len(df)
    return NewsList(
        total=total,
        page=1,
        limit=limit,
        total_pages=max(1, math.ceil(total / limit)),
        items=[_row_to_news_item(row) for _, row in featured_df.iterrows()],
    )


def get_categories(df: pd.DataFrame) -> CategoriesList:
    """Danh mục = nhóm theo nguồn báo (cột `nguon`)."""
    counts = df["nguon"].fillna("Khác").value_counts()
    items = [CategoryItem(name=str(name), count=int(count)) for name, count in counts.items()]
    return CategoriesList(total=len(items), items=items)


def search_news(
    df: pd.DataFrame, query: str, page: int = 1, limit: int = DEFAULT_PAGE_LIMIT
) -> KeywordSearchResult:
    """Tìm theo tiêu đề + tóm tắt, không phân biệt hoa/thường, không dấu."""
    q = strip_accents_lower(query.strip())

    if not q:
        matched = df.iloc[0:0]
    else:
        mask = df["_search_blob"].str.contains(q, regex=False, na=False)
        matched = df[mask]

    page_df, page, limit = _paginate_df(matched, page, limit)
    total = len(matched)
    return KeywordSearchResult(
        query=query,
        total=total,
        page=page,
        limit=limit,
        total_pages=max(1, math.ceil(total / limit)),
        items=[_row_to_news_item(row) for _, row in page_df.iterrows()],
    )

File: timkiemthongminh-main/project/frontend/src/services/newsService.ts
import type { NewsItem, NewsList, CategoriesList, SearchResult } from '../types/news'
import { apiGet, apiPost } from './apiClient'

/**
 * Service layer cho resource "news".
 *
 * Gọi trực tiếp backend FastAPI (đọc dữ liệu từ CSV, giữ trong RAM —
 * xem thư mục `backend/`). Chữ ký hàm giữ nguyên như trước để các
 * hook/component gọi chúng không cần thay đổi.
 */

export async function fetchNewsList(page = 1, limit = 10): Promise<NewsList> {
  return apiGet<NewsList>(`/api/news?page=${page}&limit=${limit}`)
}

export async function fetchCategories(): Promise<CategoriesList> {
  return apiGet<CategoriesList>('/api/categories')
}

/** Hình dạng response thô từ backend POST /search (Search Pipeline). */
interface SearchPipelineApiItem {
  id: number
  title: string | null
  summary: string | null
  content: string | null
  url: string | null
  image: string | null
  date: string | null
  source: string | null
  score: number
}

/**
 * Tìm kiếm tin tức bằng Search Pipeline (POST /search): embedding query ->
 * Qdrant search -> top kết quả -> content đầy đủ lấy từ database.
 *
 * Backend trả cố định top SEARCH_TOP_K (5), không phân trang, nên
 * `page`/`limit` được giữ lại trong chữ ký hàm để không phải sửa nơi gọi
 * (SearchPage), nhưng không còn tác dụng phân trang thực sự.
 */
export async function searchNews(
  query: string,
  _page = 1,
  limit = 10,
): Promise<SearchResult> {
  const trimmed = query.trim()
  if (!trimmed) {
    return { query, total: 0, page: 1, limit, total_pages: 1, items: [] }
  }

  const results = await apiPost<SearchPipelineApiItem[]>('/search', { query: trimmed })

  const items: NewsItem[] = results.map((r) => ({
    id: r.id,
    title: r.title,
    publish_date: r.date,
    author: null,
    source: r.source,
    url: r.url,
    comments: null,
    summary: r.summary,
    image: r.image,
  }))

  return {
    query,
    total: items.length,
    page: 1,
    limit: items.length || limit,
    total_pages: 1,
    items,
  }
}
